File: src/csharp/Intel/Iced.UnitTests/Intel/InstructionInfoTests/RegisterExtensionsTests.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if INSTR_INFO
using System;
using System.Collections.Generic;
using Iced.Intel;
using Xunit;

namespace Iced.UnitTests.Intel.InstructionInfoTests {
	public sealed class RegisterExtensionsTests {
		[Theory]
		[InlineData((Register)(-1))]
		[InlineData((Register)IcedConstants.RegisterEnumCount)]
		void GetInfo_throws_if_invalid_value(Register register) {
			Assert.Throws<ArgumentOutOfRangeException>(() => register.GetInfo());
			Assert.Throws<ArgumentOutOfRangeException>(() => register.GetBaseRegister());
			Assert.Throws<ArgumentOutOfRangeException>(() => register.GetNumber());
			Assert.Throws<ArgumentOutOfRangeException>(() => register.GetFullRegister());
			Assert.Throws<ArgumentOutOfRangeException>(() => register.GetFullRegister32());
			Assert.Throws<ArgumentOutOfRangeException>(() => register.GetSize());
		}

		[Theory]
		[MemberData(nameof(VerifyRegisterProperties_Data))]
		void VerifyRegisterProperties(Register register, int number, Register baseRegister, Register fullRegister, Register fullRegister32, int size, RegisterFlags flags) {
			var info = register.GetInfo();
			Assert.Equal(register, info.Register);
			Assert.Equal(baseRegister, info.Base);
			Assert.Equal(number, info.Number);
			Assert.Equal(fullRegister, info.FullRegister);
			Assert.Equal(fullRegister32, info.FullRegister32);
			Assert.Equal(size, info.Size);

			Assert.Equal(baseRegister, register.GetBaseRegister());
			Assert.Equal(number, register.GetNumber());
			Assert.Equal(fullRegister, register.GetFullRegister());
			Assert.Equal(fullRegister32, register.GetFullRegister32());
			Assert.Equal(size, register.GetSize());

			const RegisterFlags allFlags =
				RegisterFlags.SegmentRegister |
				RegisterFlags.GPR |
				RegisterFlags.GPR8 |
				RegisterFlags.GPR16 |
				RegisterFlags.GPR32 |
				RegisterFlags.GPR64 |
				RegisterFlags.XMM |
				RegisterFlags.YMM |
				RegisterFlags.ZMM |
				RegisterFlags.VectorRegister |
				RegisterFlags.IP |
				RegisterFlags.K |
				RegisterFlags.BND |
				RegisterFlags.CR |
				RegisterFlags.DR |
				RegisterFlags.TR |
				RegisterFlags.ST |
				RegisterFlags.MM |
				RegisterFlags.TMM;
			// If it fails, update the flags above and the code below, eg. add a IsTMM() test
			Assert.Equal(flags, flags & allFlags);

			Assert.Equal((flags & RegisterFlags.SegmentRegister) != 0, register.IsSegmentRegister());
			Assert.Equal((flags & RegisterFlags.GPR) != 0, register.IsGPR());
			Assert.Equal((flags & RegisterFlags.GPR8) != 0, register.IsGPR8());
			Assert.Equal((flags & RegisterFlags.GPR16) != 0, register.IsGPR16());
			Assert.Equal((flags & RegisterFlags.GPR32) != 0, register.IsGPR32());
			Assert.Equal((flags & RegisterFlags.GPR64) != 0, register.IsGPR64());
			Assert.Equal((flags & RegisterFlags.XMM) != 0, register.IsXMM());
			Assert.Equal((flags & RegisterFlags.YMM) != 0, register.IsYMM());
			Assert.Equal((flags & RegisterFlags.ZMM) != 0, register.IsZMM());
			Assert.Equal((flags & RegisterFlags.VectorRegister) != 0, register.IsVectorRegister());
			Assert.Equal((flags & RegisterFlags.IP) != 0, register.IsIP());
			Assert.Equal((flags & RegisterFlags.K) != 0, register.IsK());
			Assert.Equal((flags & RegisterFlags.BND) != 0, register.IsBND());
			Assert.Equal((flags & RegisterFlags.CR) != 0, register.IsCR());
			Assert.Equal((flags & RegisterFlags.DR) != 0, register.IsDR());
			Assert.Equal((flags & RegisterFlags.TR) != 0, register.IsTR());
			Assert.Equal((flags & RegisterFlags.ST) != 0, register.IsST());
			Assert.Equal((flags & RegisterFlags.MM) != 0, register.IsMM());
			Assert.Equal((flags & RegisterFlags.TMM) != 0, register.IsTMM());
		}
		public static IEnumerable<object[]> VerifyRegisterProperties_Data {
			get {
				foreach (var tc in RegisterInfoTestReader.GetTestCases())
					yield return new object[] { tc.Register, tc.Number, tc.BaseRegister, tc.FullRegister, tc.FullRegister32, tc.Size, tc.Flags };
			}
		}
	}
}
#endif

File: src/java/iced-x86/src/test/java/com/github/icedland/iced/x86/info/RegisterExtensionsTests.java
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.info;

import java.util.ArrayList;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.params.*;
import org.junit.jupiter.params.provider.*;

import com.github.icedland.iced.x86.Register;
import com.github.icedland.iced.x86.RegisterInfo;
import com.github.icedland.iced.x86.internal.IcedConstants;

final class RegisterExtensionsTests {
	@ParameterizedTest
	@MethodSource("getInfo_throws_if_invalid_value_Data")
	void GetInfo_throws_if_invalid_value(int register) {
		assertThrows(ArrayIndexOutOfBoundsException.class, () -> Register.getInfo(register));
		assertThrows(ArrayIndexOutOfBoundsException.class, () -> Register.getBaseRegister(register));
		assertThrows(ArrayIndexOutOfBoundsException.class, () -> Register.getNumber(register));
		assertThrows(ArrayIndexOutOfBoundsException.class, () -> Register.getFullRegister(register));
		assertThrows(ArrayIndexOutOfBoundsException.class, () -> Register.getFullRegister32(register));
		assertThrows(ArrayIndexOutOfBoundsException.class, () -> Register.getSize(register));
	}

	static Iterable<Arguments> getInfo_throws_if_invalid_value_Data() {
		ArrayList<Arguments> result = new ArrayList<Arguments>();
		result.add(Arguments.of(-1));
		result.add(Arguments.of(IcedConstants.REGISTER_ENUM_COUNT));
		return result;
	}

	@ParameterizedTest
	@MethodSource("verifyRegisterProperties_Data")
	void verifyRegisterProperties(int register, int number, int baseRegister, int fullRegister, int fullRegister32, int size, int flags) {
		RegisterInfo info = Register.getInfo(register);
		assertEquals(register, info.getRegister());
		assertEquals(baseRegister, info.getBase());
		assertEquals(number, info.getNumber());
		assertEquals(fullRegister, info.getFullRegister());
		assertEquals(fullRegister32, info.getFullRegister32());
		assertEquals(size, info.getSize());

		assertEquals(baseRegister, Register.getBaseRegister(register));
		assertEquals(number, Register.getNumber(register));
		assertEquals(fullRegister, Register.getFullRegister(register));
		assertEquals(fullRegister32, Register.getFullRegister32(register));
		assertEquals(size, Register.getSize(register));

		final int allFlags =
			RegisterFlags.SEGMENT_REGISTER |
			RegisterFlags.GPR |
			RegisterFlags.GPR8 |
			RegisterFlags.GPR16 |
			RegisterFlags.GPR32 |
			RegisterFlags.GPR64 |
			RegisterFlags.XMM |
			RegisterFlags.YMM |
			RegisterFlags.ZMM |
			RegisterFlags.VECTOR_REGISTER |
			RegisterFlags.IP |
			RegisterFlags.K |
			RegisterFlags.BND |
			RegisterFlags.CR |
			RegisterFlags.DR |
			RegisterFlags.TR |
			RegisterFlags.ST |
			RegisterFlags.MM |
			RegisterFlags.TMM;
		// If it fails, update the flags above and the code below, eg. add a IsTMM() test
		assertEquals(flags, flags & allFlags);

		assertEquals((flags & RegisterFlags.SEGMENT_REGISTER) != 0, Register.isSegmentRegister(register));
		assertEquals((flags & RegisterFlags.GPR) != 0, Register.isGPR(register));
		assertEquals((flags & RegisterFlags.GPR8) != 0, Register.isGPR8(register));
		assertEquals((flags & RegisterFlags.GPR16) != 0, Register.isGPR16(register));
		assertEquals((flags & RegisterFlags.GPR32) != 0, Register.isGPR32(register));
		assertEquals((flags & RegisterFlags.GPR64) != 0, Register.isGPR64(register));
		assertEquals((flags & RegisterFlags.XMM) != 0, Register.isXMM(register));
		assertEquals((flags & RegisterFlags.YMM) != 0, Register.isYMM(register));
		assertEquals((flags & RegisterFlags.ZMM) != 0, Register.isZMM(register));
		assertEquals((flags & RegisterFlags.VECTOR_REGISTER) != 0, Register.isVectorRegister(register));
		assertEquals((flags & RegisterFlags.IP) != 0, Register.isIP(register));
		assertEquals((flags & RegisterFlags.K) != 0, Register.isK(register));
		assertEquals((flags & RegisterFlags.BND) != 0, Register.isBND(register));
		assertEquals((flags & RegisterFlags.CR) != 0, Register.isCR(register));
		assertEquals((flags & RegisterFlags.DR) != 0, Register.isDR(register));
		assertEquals((flags & RegisterFlags.TR) != 0, Register.isTR(register));
		assertEquals((flags & RegisterFlags.ST) != 0, Register.isST(register));
		assertEquals((flags & RegisterFlags.MM) != 0, Register.isMM(register));
		assertEquals((flags & RegisterFlags.TMM) != 0, Register.isTMM(register));
	}

	public static Iterable<Arguments> verifyRegisterProperties_Data() {
		ArrayList<Arguments> result = new ArrayList<Arguments>();
		for (RegisterInfoTestCase tc : RegisterInfoTestReader.getTestCases())
			result.add(Arguments.of(tc.register, tc.number, tc.baseRegister, tc.fullRegister, tc.fullRegister32, tc.size, tc.flags));
		return result;
	}
}
